lib/puppet/util/windows/service.rb

```diff
@@ -741,16 +744,17 @@ def wait_on_state_transition(service, initial_state, final_state)
         state = nil
         elapsed_time = 0
         while elapsed_time <= DEFAULT_TIMEOUT
-          status = query_status(service)
-          state = status[:dwCurrentState]
-
-          return if state == final_state
-          if state == pending_state
-            Puppet.debug _("The service transitioned to the %{pending_state} state.") % { pending_state: SERVICE_STATES[pending_state] }
-            wait_on_pending_state(service, pending_state)
-            return
-          end
 
+          query_status(service) do |status|
+            state = status[:dwCurrentState]
+
+            return if state == final_state
```

Comment:
Returning from a block won't execute the clean-up code -- this should return `next` instead.

Reply:
I would have hoped that FFI would have wrapped any cleanup in ensure blocks to avoid that, but TBH I don't trust FFI that much so I think you're right.

Reply:
See general comment about this: clean up of pointers is done in garbage collection (rather than in an ensure or cleanup after the block), so whatever we do here will be fine.